File: browser/typescript/lib/src/htmlApp/player.ts
import { StreamIds } from "../rpc/index";
import { handlerListenners, rpc } from "./rpc_call";
import { fncCall, getFncCall, getSetCall, newHandlerName, } from "./handlers";

export class PlayerProxy {
    static async create(url: string, listen?: (streamId: StreamIds, value: any)=>void) {
        let res = new PlayerProxy();
        if (listen) handlerListenners[res.audioName] = listen;
        await fncCall(null, 'createPlayer', [res.playerName, res.audioName, url])
        return res
    }
    playerName = newHandlerName();
    audioName = newHandlerName();
    async dispose() {
        await fncCall(this.playerName, 'dispose')
        delete handlerListenners[this.audioName]
    }
    play() {
        return fncCall(this.audioName, 'play')
    }
    stop() {
        return rpc([
            getFncCall(this.audioName, 'pause'),
            getSetCall(this.audioName, 'currentTime', 0),
        ])
    }
}

File: packages/webview_lib/lib/src/player/player.dart
import '../rpc/handlers.dart';
import '../rpc/rpc.dart';

export 'interface.dart';

class PlayerProxy {
  static Future<PlayerProxy> create(String url, {void Function(int streamId, IRpcResult value)? listen}) async {
    final res = new PlayerProxy();
    handlerListenners[res.audioName] = (streamId, value) {
      switch (streamId) {
        case StreamIds.playerError:
          throw Exception(value.toString());
        default:
          listen?.call(streamId, value);
          break;
      }
    };
    await fncCall(null, 'createPlayer', [res.playerName, res.audioName, url]);
    return res;
  }

  final playerName = newHandlerName();
  final audioName = newHandlerName();
  Future dispose() async {
    await fncCall(this.playerName, 'dispose');
    handlerListenners.remove(audioName);
  }

  Future play() => fncCall(this.audioName, 'play');
  Future stop() => rpc([
        getFncCall(this.audioName, 'pause'),
        getSetCall(this.audioName, 'currentTime', 0),
      ]);
}
